pkg/pillar/cas/containerd.go

```diff
@@ -0,0 +1,407 @@
+package cas
+
+import (
+	"context"
+	"fmt"
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/leases"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/containerd/containerd/snapshots"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/opencontainers/go-digest"
+	"github.com/opencontainers/image-spec/identity"
+	spec "github.com/opencontainers/image-spec/specs-go/v1"
+	"runtime"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+	"io"
+	"time"
+)
+
+var (
+	ctrdClient   *containerd.Client
+	ctrdCtx      context.Context
+	contentStore content.Store
+)
+
+const (
+	// containerd socket
+	ctrdSocket = "/run/containerd/containerd.sock"
+	// ctrdServicesNamespace containerd namespace for running containers
+	ctrdServicesNamespace = "eve-user-apps"
+	//containerdRunTime - default runtime of containerd
+	containerdRunTime = "io.containerd.runtime.v1.linux"
+	// default snapshotter used by containerd
+	defaultSnapshotter = "overlayfs"
+)
```

Comment:
A lot of this is plain copy of things from containerd/containerd.go -- I think we need to reconcile it and move it into the right place.

My suggestion is to:
    * split this into 3 interfaces CAS, Image and Snapshot
    * provide interface implementations from under containerd/containerd.go

What do you think @adarsh-zededa @deitch ?

Reply:
See the [above comment](https://github.com/lf-edge/eve/pull/919#discussion_r417898529).

For now, I think we need to leave it as one interface that handles it all. We always can refactor later, if we can find a clean way to do it. In any case, I suspect the API will change somewhat as we do implementation and discover some use case or side effect we missed. I would not waste any more time making the abstract perfect, just move it ahead and iterate.

For the other point - duplication - I think you are right. Let's get this in as is, and then immediately start a PR that gets rid of most of `cas/containerd.go`. We have two options:

* add functions to `pkg/pillar/containerd/` so that it has a struct that implements the `CAS` interface
* keep a light wrapper in `pkg/pillar/cas/containerd.go` that just calls the functions out of `pkg/pillar/containerd/`

Either approach is fine. @adarsh-zededa can do whatever he finds easiest, methinks. Merge this in first, and then do that.